src/Storage/Notmuch.hs

```diff
@@ -40,6 +40,22 @@ getMessages s settings =
               mails <- liftIO $ mapM messageToMail msgs
               pure $ Vec.fromList mails
 
+-- | creates a vector of parsed mails which are tips of each thread
+--
+getTopMessages
+  :: (HasThreads Query, MonadError Error m, MonadIO m)
+  => T.Text
+  -> NotmuchSettings FilePath
+  -> m (Vec.Vector NotmuchMail)
+getTopMessages s settings =
+  bracketT (databaseOpenReadOnly (view nmDatabase settings)) databaseDestroy go
+  where
+    go db = do
+        tds <- query db (FreeForm $ T.unpack s) >>= threads >>= mapM threadToplevelMessages
```

Comment:
Instead of `mapM` you can often use `traverse` which uses the weaker `Applicative` abstraction.
I think you probably can use `traverse` in this case.

Reply:
Good to know that instead of `mapM` rather use `traverse`... changed this.